system/src/system_power_manager.cpp

```diff
@@ -357,22 +369,57 @@ void PowerManager::isrHandler() {
 void PowerManager::initDefault(bool dpdm) {
   PMIC power(true);
   power.begin();
+
+#if HAL_PLATFORM_POWER_MANAGEMENT_PMIC_WATCHDOG
+  // We keep the watchdog running at 40s and make sure to kick it periodically
+  // FIXME: Make sure to adjust PMIC_WATCHDOG_INTERVAL accordingly
+  power.resetWatchdog();
+  power.setWatchdog(0b01);
+#else
   // Enters host-managed mode
   power.disableWatchdog();
+#endif // HAL_PLATFORM_POWER_MANAGEMENT_PMIC_WATCHDOG
 
   // Adjust charge voltage
   power.setChargeVoltage(config_.termination_voltage);
 
   // Set recharge threshold to default value - 100mV
   power.setRechargeThreshold(100);
   power.setChargeCurrent(config_.charge_current);
-  if (dpdm) {
-    // Force-start input current limit detection
-    power.enableDPDM();
-  }
+
   // Enable charging
   power.enableCharging();
-  power.enableBuck();
+
+  // Just in case make sure to enable BATFET
+  if (!power.isBATFETEnabled()) {
+    power.enableBATFET();
+  }
+
+  // Enable buck converter
+  bool inDpDm = power.isInDPDM();
+  if (!power.isBuckEnabled()) {
+    power.enableBuck();
+    // Make sure we restart DPDM as otherwise we may get stuck
+    // with an invalid ILIM due to reading back and writing
+    // a transient value by modifying the same register with enableBuck()
+    if (inDpDm) {
+      dpdm = true;
+    }
+  }
+
+  auto powerSource = powerSourceFromStatus(power.getSystemStatus());
+  bool vinWithValidLimit = (power.getInputCurrentLimit() > 100) &&
+      ((powerSource == POWER_SOURCE_VIN) ||
+      ((config_.flags & HAL_POWER_USE_VIN_SETTINGS_WITH_USB_HOST) && powerSource == POWER_SOURCE_USB_HOST));
+
+  if (dpdm) {
+    if (inDpDm || !vinWithValidLimit) {
+      // Force-start input current limit detection
+      // only if we modified REG00 while in DPDM or we are not powered by VIN
+      // and we have a non-100mA limit set
+      power.enableDPDM();
```

Comment:
Will this actually restart the detection if REG07[7] is already 1?

Reply:
👍 Let's double check. From what I've seen - yes.

Reply:
Doesn't work. A change has been added to wait for it to finish.